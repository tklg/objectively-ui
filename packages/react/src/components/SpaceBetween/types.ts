import { ForwardRefExoticComponent, ReactElement } from 'react'
import { CommonSize } from 'src/types/sizes'

export interface SpaceBetweenProps {
  children: ReactElement | ReactElement[];
  className?: string;
  direction?: 'horizontal' | 'vertical';
  size?: CommonSize;
}

export type SpaceBetweenComponent<P = SpaceBetweenProps> = ForwardRefExoticComponent<P>
