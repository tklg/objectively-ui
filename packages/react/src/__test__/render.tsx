import { CssVarsProvider } from 'src/providers'
import { render, RenderOptions } from '@testing-library/react'
import { FC, PropsWithChildren, ReactElement } from 'react'

const AllTheProviders: FC<PropsWithChildren> = ({children}) => {
  return <CssVarsProvider>{children}</CssVarsProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
