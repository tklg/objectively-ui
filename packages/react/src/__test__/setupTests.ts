import '@testing-library/jest-dom/extend-expect'
import matchMediaMock from 'src/__test__/matchMedia.mock'
import { createSerializer } from '@emotion/jest'

matchMediaMock()

// configures @emotion/jest to not insert styles
expect.addSnapshotSerializer(createSerializer({
  includeStyles: false,
}))
