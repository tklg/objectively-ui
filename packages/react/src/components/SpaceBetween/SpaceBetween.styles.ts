import { css } from '@emotion/react'
import { ColorTheme } from 'src/types/ColorTheme'
import { CommonSize } from 'src/types/sizes'

export const spaceBetweenStyles = (theme: ColorTheme, size: CommonSize) => css({
  display: 'inline-flex',

  '&[data-spacebetween-direction=horizontal]': {
    flexDirection: 'row',
  },
  '&[data-spacebetween-direction=vertical]': {
    flexDirection: 'column',
  },
  '&[data-spacebetween-direction=horizontal] > *:not([data-spacebetween-first=true])': {
    marginLeft: theme.spacing[size],
  },
  '&[data-spacebetween-direction=vertical] > *:not([data-spacebetween-first=true])': {
    marginTop: theme.spacing[size],
  },
})
