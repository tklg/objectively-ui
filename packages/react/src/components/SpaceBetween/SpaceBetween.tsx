import { Children, cloneElement, forwardRef, ReactElement } from 'react'
import { spaceBetweenStyles } from 'src/components/SpaceBetween/SpaceBetween.styles'
import { SpaceBetweenComponent, SpaceBetweenProps } from 'src/components/SpaceBetween/types'
import { useTheme } from 'src/hooks'
import { buildClassName } from 'src/utils/buildClassName'

const ELEMENT_NAME = 'SpaceBetween'

export const SpaceBetween = forwardRef<HTMLDivElement, SpaceBetweenProps>(({
  children,
  direction = 'horizontal',
  size = 'sm',
  className: _className,
  ...props
}, ref) => {
  const theme = useTheme()
  const className = buildClassName(ELEMENT_NAME, {
    size,
    direction,
  }, _className)

  return (
    <div
      ref={ref}
      {...props}
      className={className}
      css={spaceBetweenStyles(theme, size)}
      data-spacebetween-direction={direction}
    >
      {Children.toArray(children).map((elem, i) => (
        typeof elem === 'object'
          ? cloneElement(elem as ReactElement, {
            key: i,
            'data-spacebetween-first': i === 0 || undefined,
          })
          : <span key={i} data-spacebetween-first={i === 0 || undefined}>{elem}</span>
      ))}
    </div>
  )
}) as SpaceBetweenComponent

if (process.env.NODE_ENV !== 'production') {
  SpaceBetween.displayName = ELEMENT_NAME
}
