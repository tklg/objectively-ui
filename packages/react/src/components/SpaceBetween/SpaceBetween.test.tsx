/* eslint-disable @typescript-eslint/ban-ts-comment */
import { SpaceBetween } from 'src/components/SpaceBetween/SpaceBetween'
import { render, screen } from 'src/__test__/render'

describe('SpaceBetween', () => {
  it('renders a single item', async () => {
    render(
      <SpaceBetween>
        <button>text 1</button>
      </SpaceBetween>,
    )

    expect(screen.getByText('text 1')).toHaveAttribute('data-spacebetween-first')
  })

  it('renders multiple items', async () => {
    render(
      <SpaceBetween>
        <button>text 1</button>
        <button>text 2</button>
      </SpaceBetween>,
    )

    expect(screen.getByText('text 1')).toHaveAttribute('data-spacebetween-first')
    expect(screen.getByText('text 2')).not.toHaveAttribute('data-spacebetween-first')
  })

  it('wraps items in a span if they are not a component', async () => {
    // @ts-ignore
    render(<SpaceBetween data-testid='test'>text 1<button>text 2</button></SpaceBetween>)

    expect(screen.getByText('text 1')).toHaveAttribute('data-spacebetween-first')
    expect(screen.getByText('text 2')).not.toHaveAttribute('data-spacebetween-first')
    expect(screen.getByTestId('test')).toMatchSnapshot()
  })
})
