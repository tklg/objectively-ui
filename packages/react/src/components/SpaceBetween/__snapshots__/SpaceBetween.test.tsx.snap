// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpaceBetween wraps items in a span if they are not a component 1`] = `
<div
  class="ObjUI-SpaceBetween ObjUI-SpaceBetween-sizeSm ObjUI-SpaceBetween-directionHorizontal emotion-0"
  data-spacebetween-direction="horizontal"
  data-testid="test"
>
  <span
    data-spacebetween-first="true"
  >
    text 1
  </span>
  <button>
    text 2
  </button>
</div>
`;
